File: src/y2023/Day08.java
package y2023;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;


public class Day08 {
    public static void main(String[] args) throws IOException {
        List<String> testInput = readInput("Day08_test_part1");
        System.out.println(part1(testInput));
//        System.out.println(part2(testInput));

        List<String> input = readInput("Day08");
        System.out.println(part1(input));
//        System.out.println(part2(input));
    }

    private static List<String> readInput(String fileName) throws IOException {
        return Files.readAllLines(Paths.get(fileName));
    }

    private static String pickDirection(Pair<String, String> node, char dir) {
        return 'L' == dir ? node.left : node.right;
    }

    private static int part1(List<String> input) {
        String pattern = input.get(0);
        Iterator<Character> ruleSeq = new Iterator<>() {

            private int index = 0;
            @Override
            public boolean hasNext() {
                return true;
            }

            @Override
            public Character next() {
                char res = pattern.charAt(index);
                index = (index + 1) % pattern.length();
                return res;
            }
        };
        Map<String, Pair<String, String>> map = new HashMap<>();
        input.stream().skip(2).filter(s -> !s.isBlank()).forEach(s ->{
            String key = s.substring(0, s.indexOf(" ="));
            String[] parts = s.substring(s.indexOf("= (") + 3, s.indexOf(")")).split(", ");
            map.put(key, new Pair<>(parts[0], parts[1]));
        });

        String node = "AAA";
        int count = 0;
        while (!"ZZZ".equals(node)) {
            node = pickDirection(map.get(node), ruleSeq.next());
            count++;
        }
        return count;
    }

    record Pair<L, R>(L left, R right) { }
}

File: src/y2023/Day08.kt
package y2023

import readInput


fun main() {

    fun Pair<String, String>.pickDirection(dir: Char): String {
        return if ('L' == dir) this.first else this.second
    }

    fun part1(input: List<String>): Int {
        val pattern = input.first()
        val ruleSeq = generateSequence(0) { (it + 1) % pattern.length }.map { pattern[it] }
        val map = input.drop(2).filter { it.isNotBlank() }
            .associate {
                val key = it.substringBefore(" =")
                val (left, right) = it.substringAfter("= (").substringBefore(")").split(", ")
                key to Pair(left, right)
            }
        return ruleSeq.runningFold("AAA") { node, rule ->
            map[node]!!.pickDirection(rule)
        }.takeWhile { it != "ZZZ" }.count()
    }

    fun gcd(a: Long, b: Long): Long {
        if (b == 0L) return a
        return gcd(b, a % b)
    }

    fun part2(input: List<String>): Int {
        val pattern = input.first()
        val ruleSeq = generateSequence(0) { (it + 1) % pattern.length }.map { pattern[it] }
        val map = input.drop(2).filter { it.isNotBlank() }
            .associate {
                val key = it.substringBefore(" =")
                val (left, right) = it.substringAfter("= (").substringBefore(")").split(", ")
                key to Pair(left, right)
            }
        val startingNodes = map.filter { it.key.endsWith("A") }.keys.toSet()
        val map1 = startingNodes.map { startingNode ->
            ruleSeq.runningFold(startingNode) { currentNode, rule ->
                println("$startingNode -> $rule")
                map[currentNode]!!.pickDirection(rule)
            }.takeWhile { !it.endsWith("Z") }.count()
        }
        // this is the most important part
        val reduce = map1.map { it.toLong() }.reduce { a, b ->
            println("$a (a) * $b (b) / ${gcd(a,b)} (gcd)")
            a * b / gcd(a, b)
        }
        println(reduce)
        return 0
    }


    // test if implementation meets criteria from the description, like:
    var testInput = readInput("y2023", "Day08_test_part1")
//    println(part1(testInput))
//    check(part1(testInput) == 6440L)
    println(part2(testInput))
//    check(part2(testInput) == 251927063)

    val input = readInput("y2023", "Day08")
    println(part1(input))
//    check(part1(input) == 251927063L)
    println(part2(input))
//    check(part2(input) == 255632664)
}